### PR TITLE
Dashboard Updates

### PR DESCRIPTION
**Pull request by eslachance**
Enmap updated to version 4
"set" command updated to use enmap 4
"message" event updated to use enmap 4
"setup.js" updated to use readline-sync
Dashboard updated to use memorycache rather than level-session-store
enmap-level dependency removed
enquire dependency (and 26 sub-deps) removed
readline-sync dependency added (no sub-deps)
level-session-store dependency removed

### CODE REVIEW

File: config.js.example [hunk 1]
```diff
@@ -1,6 +1,6 @@
 const config = {
-  // Bot Owner, level 10 by default. You no longer need to supply the owner ID, as the bot
-  // will pull this information directly from it's application page.
+  // Bot Owner, level 10 by. You do not need to supply the owner ID, as the bot
```
Comment (YorkAARGH) [inline config.js.example]: You accidentally removed the word "default".

File: dashboard/templates/blocks/header.ejs [hunk 2]
```diff
@@ -12,18 +12,18 @@
     <meta property="og:site_name" content="<%= bot.user.username %> Dashboard">
     <meta property="og:locale" content="en_US">
 
-    <link rel="stylesheet" href="/public/css/bootstrap.min.css">
-    <link rel="stylesheet" href="/public/css/font-awesome.min.css">
-    <link rel="stylesheet" href="/public/css/darkly.css">
-    <link rel="stylesheet" href="/public/css/dashboard.css" />
-    <link rel='icon' href='<%= bot.user.avatarURL %>'>
-
+    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
```
Comment (YorkAARGH) [inline dashboard/templates/blocks/header.ejs]: So you're directly linking to their CDN again? what if it changes? That's why we included the files.

File: index.js [hunk 3]
```diff
@@ -138,17 +135,64 @@ class GuideBot extends Discord.Client {
...
+
+  /*
+  MESSAGE CLEAN FUNCTION
+
+  "Clean" removes @everyone pings, as well as tokens, and makes code blocks
+  escaped so they're shown more easily. As a bonus it resolves promises
+  and stringifies objects!
+  This is mostly only used by the Eval and Exec commands.
+  */
+  async clean (client, text) {
```
Comment (YorkAARGH) [inline index.js]: This function has a space between the function name and the parenthesis, but like 172 doesn't, any chance you could fix this for consistency?

Comment (MrJacz) [reply index.js]: why does it even pass client, its in the client 🤔

File: commands/set.js [hunk 4]
```diff
@@ -30,33 +30,42 @@ class Set extends Command {
...
+      // User must specify a key.
       if (!key) return message.reply("Please specify a key to edit");
+      // User must specify a key that actually exists!
       if (!settings[key]) return message.reply("This key does not exist in the settings");
+      // User must specify a value to change.
       if (value.length < 1) return message.reply("Please specify a new value");
-    
-      settings[key] = value.join(" ");
+      // User must specify a different value than the current one.
+      if (value.join(" ") === settings[key]) return message.reply("This setting already has that value!");
```
Comment (MrJacz) [inline commands/set.js]: `value.join(" ")` is done more then once, make a variable and use that

File: util/dashboard.js [hunk 5]
```diff
@@ -98,7 +98,7 @@ module.exports = (client) => {
   // Session data, used for temporary storage of your visitor's session information.
   // the `secret` is in fact a "salt" for the data, and should not be shared publicly.
   app.use(session({
-    store: new LevelStore("./data/dashboard-session/"),
+    store: new MemoryStore({ checkPeriod: 86400000 }),
```
Comment (MrJacz) [inline util/dashboard.js]: session defaults to MemoryStore, iirc

File: index.js [hunk 6]
```diff
@@ -138,17 +135,64 @@ class GuideBot extends Discord.Client {
...
+    } catch (e) {
+      return false;
+    }
+  };
+
+  /* 
+  WAIT FUNCTION
+  Basically just an async shortcut to using a setTimeout. Nothing fancy!
+  */
+  wait() {
```
Comment (bdistin) [inline index.js]: You might want to change how this is done so you don't have to do `this.client.wait()(time)`, or re-promisify setTimout every time you want to wait for some time.

You probably want `this.wait = promisify(setTimeout);` in the constructor, so you can do `await this.client.wait(time);` and the like.

File: util/setup.js [hunk 7]
```diff
@@ -14,81 +13,55 @@ const defaultSettings = `{
...
+  } else {
     console.log("First Start! Inserting default guild settings in the database...");
-    await settings.setAsync("default", defaultSettings);
+    settings.set("default", defaultSettings);
   }
-  const isGlitch = await inquirer.prompt([{type: "confirm", name: "glitch", message: "Are you hosted on Glitch.com?", default: false}]);
-
-  if (isGlitch.glitch) {
+  const isGlitch = input.keyInYN("Are you hosted on Glitch.com? ");
+
```
Comment (bdistin) [inline util/setup.js]: Seems there is a trailing space here?

File: config.js.example (hunk 1 above)
Comment (eslachance) [reply config.js.example]: No I forgot to remove the `by`, the level 10 is pretty much set in stone

File: dashboard/templates/blocks/header.ejs (hunk 2 above)
Comment (eslachance) [reply dashboard/templates/blocks/header.ejs]: The change here is that I'm pointing towards the real CDNs for the "main" files, but not for the custom theme and custom CSS. It's much better to CDN things that are common. This ain't the shitty boostraptheme website that keeps changing.

File: index.js (hunk 3 above)
Comment (eslachance) [reply index.js]: Fixed, for both comments.

File: commands/set.js (hunk 4 above)
Comment (eslachance) [reply commands/set.js]: Fixed

File: util/dashboard.js (hunk 5 above)
Comment (eslachance) [reply util/dashboard.js]: The built-in memory store is shit though, according to that module's author. it has a memory leak, that's why.

File: index.js (hunk 6 above)
Comment (eslachance) [reply index.js]: Fixed

File: config.js.example (hunk 1 above)
Comment (eslachance) [reply config.js.example]: Then again. Consistency. Put "default" back.

File: util/setup.js (hunk 7 above)
Comment (eslachance) [reply util/setup.js]: Fixed

**Pull request merged**